File: ios/TerritoryRun/Networking/ApiClient.swift
import Foundation

struct GpsPoint: Codable {
    let lat: Double
    let lon: Double
    let ts: Int64
    let acc: Double
    let speed: Double
}

struct UserProfile: Codable {
    let user_id: String
    let username: String
    let display_name: String
    let cells_owned: Int
    let territories_captured: Int
    let total_capture_score: Int
    let activity_count: Int
}

struct CompleteActivityResponse: Codable {
    let activity_id: String
    let status: String
    let capture_result: CaptureResult?
    let trust_score: Int
}

struct CaptureResult: Codable {
    let new_cells: Int
    let capture_score: Int
    let h3_cells: [String]
}

enum ApiClient {
    static var token: String?

    static func me() async throws -> UserProfile {
        try await request(path: "/api/v1/users/me", method: "GET")
    }

    static func createActivity(idempotencyKey: String) async throws -> String {
        struct Body: Codable { let started_at: String; let idempotency_key: String }
        struct Resp: Codable { let activity_id: String }
        let resp: Resp = try await request(
            path: "/api/v1/activities",
            method: "POST",
            body: Body(started_at: ISO8601DateFormatter().string(from: Date()), idempotency_key: idempotencyKey)
        )
        return resp.activity_id
    }

    static func uploadPoints(activityId: String, points: [GpsPoint]) async throws {
        struct Body: Codable { let points: [GpsPoint] }
        try await requestVoid(path: "/api/v1/activities/\(activityId)/points", method: "POST", body: Body(points: points))
    }

    static func completeActivity(activityId: String, distanceM: Double, durationS: Int, cells: [String]) async throws -> CompleteActivityResponse {
        struct Body: Codable {
            let ended_at: String
            let distance_m: Double
            let duration_s: Int
            let h3_cells_hint: [String]
        }
        return try await request(
            path: "/api/v1/activities/\(activityId)/complete",
            method: "POST",
            body: Body(
                ended_at: ISO8601DateFormatter().string(from: Date()),
                distance_m: distanceM,
                duration_s: durationS,
                h3_cells_hint: cells
            )
        )
    }

    private static func request<T: Decodable>(path: String, method: String, body: Encodable? = nil) async throws -> T {
        let url = URL(string: AppConfig.apiBaseURL + path)!
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let body {
            req.httpBody = try JSONEncoder().encode(AnyEncodable(body))
        }
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard (resp as? HTTPURLResponse)?.statusCode ?? 500 < 300 else {
            throw URLError(.badServerResponse)
        }
        return try JSONDecoder().decode(T.self, from: data)
    }

    private static func requestVoid(path: String, method: String, body: Encodable) async throws {
        let _: [String: String] = try await request(path: path, method: method, body: body)
    }
}

private struct AnyEncodable: Encodable {
    let value: Encodable
    init(_ value: Encodable) { self.value = value }
    func encode(to encoder: Encoder) throws { try value.encode(to: encoder) }
}

File: android/app/src/main/java/com/territoryrun/app/data/remote/ApiClient.kt
package com.territoryrun.app.data.remote

import com.territoryrun.app.BuildConfig
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import java.util.concurrent.TimeUnit

data class CreateActivityBody(
    val started_at: String,
    val idempotency_key: String,
    val device_info: Map<String, String> = emptyMap()
)

data class CreateActivityResponse(val activity_id: String)

data class GpsPointDto(
    val lat: Double,
    val lon: Double,
    val ts: Long,
    val acc: Double,
    val speed: Double = 0.0
)

data class UploadPointsBody(val points: List<GpsPointDto>)

data class CompleteActivityBody(
    val ended_at: String,
    val distance_m: Double,
    val duration_s: Int,
    val h3_cells_hint: List<String> = emptyList()
)

data class CaptureResultDto(
    val new_cells: Int,
    val total_cells: Int,
    val capture_score: Int,
    val h3_cells: List<String>
)

data class CompleteActivityResponse(
    val activity_id: String,
    val status: String,
    val capture_result: CaptureResultDto?,
    val trust_score: Int,
    val message: String?
)

data class UserProfileDto(
    val user_id: String,
    val username: String,
    val display_name: String,
    val cells_owned: Int,
    val territories_captured: Int,
    val total_capture_score: Int,
    val activity_count: Int
)

interface TerritoryApi {
    @POST("api/v1/activities")
    suspend fun createActivity(@Body body: CreateActivityBody): CreateActivityResponse

    @POST("api/v1/activities/{id}/points")
    suspend fun uploadPoints(@Path("id") id: String, @Body body: UploadPointsBody): Map<String, Any>

    @POST("api/v1/activities/{id}/complete")
    suspend fun completeActivity(@Path("id") id: String, @Body body: CompleteActivityBody): CompleteActivityResponse

    @GET("api/v1/users/me")
    suspend fun me(): UserProfileDto
}

object ApiClient {
  private var tokenProvider: () -> String = { "" }

  fun setTokenProvider(provider: () -> String) {
    tokenProvider = provider
  }

  private val authInterceptor = Interceptor { chain ->
    val token = tokenProvider()
    val request = if (token.isNotBlank()) {
      chain.request().newBuilder().addHeader("Authorization", "Bearer $token").build()
    } else chain.request()
    chain.proceed(request)
  }

  private val client = OkHttpClient.Builder()
    .connectTimeout(30, TimeUnit.SECONDS)
    .readTimeout(60, TimeUnit.SECONDS)
    .addInterceptor(authInterceptor)
    .addInterceptor(HttpLoggingInterceptor().apply { level = HttpLoggingInterceptor.Level.BODY })
    .build()

  val api: TerritoryApi = Retrofit.Builder()
    .baseUrl(ensureTrailingSlash(BuildConfig.API_BASE_URL))
    .client(client)
    .addConverterFactory(GsonConverterFactory.create())
    .build()
    .create(TerritoryApi::class.java)

  private fun ensureTrailingSlash(url: String): String =
    if (url.endsWith("/")) url else "$url/"
}
